hcl/writer.go

```diff
@@ -128,30 +129,72 @@ func (w *Writer) Sync() error {
-	// first loop: the resource types (e.g aws_security_group)
-	for rt, body := range resources.(map[string]map[string]interface{}) {
-		// second loop: the resource names (e.g "YefiF")
-		for name, resource := range body {
-			b, err := json.Marshal(resource)
-			if err != nil {
-				return errors.Wrap(err, "unable to convert resources to JSON")
-			}
-			for key, value := range i {
-				// if the value of the interpolation is ${resource_type.name.id}
-				// and we are inside the component `resource_type.name` we continue
-				// to the next iteration
-				if strings.Contains(value, rt) {
-					continue
-				}
-				b = bytes.ReplaceAll(b, []byte(key), []byte(value))
-			}
-			var interpolatedResource map[string]interface{}
-			if err := json.Unmarshal(b, &interpolatedResource); err != nil {
-				return errors.Wrap(err, "unable to convert JSON to resource array")
-			}
-			resources.(map[string]map[string]interface{})[rt][name] = interpolatedResource
+	// we need to isolate each resource
+	// getting each resource is easier to avoid cycle
+	// or interpolaception.
+	// We first loop over resource type (e.g: aws_instance)
+	for rt, resource := range resources.(map[string]map[string]interface{}) {
+		// we loop over a resource (e.g: aws_instance.oDSOj)
+		for name, block := range resource {
+			src := reflect.ValueOf(block)
+			// this will store the updated block
+			dest := reflect.New(src.Type()).Elem()
+			// walk through the resources to interpolate the good values
+			walk(dest, src, i, name, rt)
+			// remove reflect.Value wrapper from dest
+			resources.(map[string]map[string]interface{})[rt][name] = dest.Interface()
 		}
 	}
 	return nil
 }
+
+// walk through a resource block. it's easier since we do not how the block is made
+func walk(dest, src reflect.Value, interpolate map[string]string, name string, resourceType string) {
+	switch src.Kind() {
+	// it's an interface, so we basically need
+	// to extract the elem and walk through it
+	// as the initial call
+	case reflect.Interface:
+		srcValue := src.Elem()
+		destValue := reflect.New(srcValue.Type()).Elem()
+		walk(destValue, srcValue, interpolate, name, resourceType)
+		dest.Set(destValue)
+
+	// if the current `src` is a slice
+	// we iterate on each element.
+	case reflect.Array, reflect.Slice:
+		dest.Set(reflect.MakeSlice(src.Type(), src.Len(), src.Cap()))
+		for i := 0; i < src.Len(); i++ {
+			walk(dest.Index(i), src.Index(i), interpolate, name, resourceType)
+		}
+
+	// it's a map
+	case reflect.Map:
+		dest.Set(reflect.MakeMap(src.Type()))
+		iter := src.MapRange()
+		for iter.Next() {
+			// New gives us a pointer, but again we want the value
+			destValue := reflect.New(iter.Value().Type()).Elem()
+			walk(destValue, iter.Value(), interpolate, name, resourceType)
+			dest.SetMapIndex(iter.Key(), destValue)
+		}
+
+	// what we want to interpolate is a string
+	// we do not interpolate a custom tag (like cycloid.io) since it's key.
+	case reflect.String:
```

Comment:
Commented on PM with you already but to keep the record, could it be anything else than String no? Int, bool etc.